browser/ui/ai_chat/print_preview_extractor.cc

```diff
@@ -0,0 +1,562 @@
+      dict.Set(printing::kSettingPreviewModifiable, !is_pdf_);
+      auto url = web_contents_->GetLastCommittedURL();
+      dict.Set(printing::kSettingHeaderFooterURL, url.spec());
+      OnPrintPreviewRequest(preview_request_id_);
+      print_render_frame_->PrintPreview(std::move(dict));
+    }
+  }
+
+ private:
+  void SendResult(std::string result) {
+    PreviewCleanup();
+    std::move(callback_).Run(std::move(result));
+  }
+
+  // printing::mojom::PrintPreviewUI:
+  void SetOptionsFromDocument(
+      const printing::mojom::OptionsFromDocumentParamsPtr params,
+      int32_t request_id) override {}
+
+  void DidPrepareDocumentForPreview(int32_t document_cookie,
+                                    int32_t request_id) override {
+    DVLOG(3) << __func__ << ": id=" << request_id;
+    if (is_pdf_) {
+      return;
+    }
+    // For case of print preview, page metafile is used to composite into
+    // the document PDF at same time.  Need to indicate that this scenario
+    // is at play for the compositor.
+    auto* client = PrintCompositeClient::FromWebContents(web_contents_);
+    DCHECK(client);
+    if (client->GetIsDocumentConcurrentlyComposited(document_cookie)) {
+      return;
+    }
+
+    content::RenderFrameHost* render_frame_host =
+        printing::GetFrameToPrint(web_contents_.get());
+    // |render_frame_host| could be null when the print preview dialog is
+    // closed.
+    if (!render_frame_host) {
+      return;
+    }
+
+    client->PrepareToCompositeDocument(
+        document_cookie, render_frame_host,
+        printing::GetCompositorDocumentType(),
+        mojo::WrapCallbackWithDefaultInvokeIfNotRun(
+            base::BindOnce(
+                &PrintPreviewExtractorInternal::OnPrepareForDocumentToPdfDone,
+                weak_ptr_factory_.GetWeakPtr(), request_id),
+            printing::mojom::PrintCompositor::Status::kCompositingFailure));
+  }
+
+  void DidPreviewPage(printing::mojom::DidPreviewPageParamsPtr params,
+                      int32_t request_id) override {
+    DVLOG(3) << __func__ << ": id=" << request_id;
+    uint32_t page_index = params->page_index;
+    const printing::mojom::DidPrintContentParams& content = *params->content;
+    if (page_index == printing::kInvalidPageIndex ||
+        !content.metafile_data_region.IsValid()) {
+      return;
+    }
+    if (is_pdf_) {
+      OnCompositePdfPageDone(page_index, params->document_cookie, request_id,
+                             printing::mojom::PrintCompositor::Status::kSuccess,
+                             content.metafile_data_region.Duplicate());
+      return;
+    }
+
+    auto* client = PrintCompositeClient::FromWebContents(web_contents_);
+    DCHECK(client);
+
+    content::RenderFrameHost* render_frame_host =
+        printing::GetFrameToPrint(web_contents_.get());
+    if (!render_frame_host) {
+      DLOG(ERROR) << "No render frame host for print preview";
+      SendResult("");
```

Comment:
Well in current implementation it also early returns but yes it doesn't send the result. I now see that this is per-page. What do you recommend? Is this not fatal if it happens on every page? Could it happen?

Reply:
Early return doesn't mean we can sending result. just remove the sending result back.
This is unlikely to happen. Even it happens on all pages, we should handle it in `MetafileReadyForPrinting` which is called when `PrintRenderFrameHelper::FinalizePrintReadyDocument` finishes

Reply:
The effect of early return here is we just don't send the page into compositor then it would match `printing::mojom::DidPreviewDocumentParamsPtr->expected_pages_count` when one or more pages went wrong when calling `FinishDocumentComposition`. So nothing is fatal here.

Reply:
This check was originally extracted from `//chrome/browser/ui/webui/print_preview/print_preview_ui.cc` but the difference would be we don't have a print preview dialog which they check its rfh here so it is possible for `print_preview_ui.cc` to have empty rfh when dialog closed. But for our use case that won't happen, the only way it can be nullptr is tab closed.

Reply:
So early return is good enough and the original callback will still eventually get fired? 8c401ad7b3bf8f2df987571698d4e0e4fc9256d2

Reply:
yes